### PR TITLE
Add  `rename_all` attribute to `#[pyclass]`

### LINKED ISSUE #3383: Support renaming enum variants
It would be really convenient if the `#[pyclass]` attribute offered a way to rename all variants of an enum.

## The problem

By default the generated Python enums expose variants named after Rust's naming convention, which is not idiomatic for Python where SCREAMING_SNAKE_CASE is the norm. One can either create a new enum specifically tailored for PyO3, or apply `#[pyo3(name = "...")]` to every variants. Neither of these options are practical for big enums, and I can't make the second one work if I need to conditionally support PyO3: `#[cfg_attr(feature = "pyo3", pyo3(name = "...")]` won't compile.

## Solution

Add a new argument to `#[pyclass]` that would only apply to enums and that would rename all their variants: `rename_all` for instance.

It could be a good idea to also support `#[serde(rename_all = "...")]` if it is already applied to an enum, although I don't think it would happen often that people use screaming snake case when (de)serializing their types.

### PR DESCRIPTION
**Pull request by DataTriny**
Closes #3383

### CODE REVIEW

File: pyo3-macros-backend/src/pyclass.rs
Comment (adamreichold) [inline pyo3-macros-backend/src/pyclass.rs:215]: Couldn't this just as well be applied to generated getters and setters?

File: pyo3-macros-backend/src/pyclass.rs
Comment (adamreichold) [inline pyo3-macros-backend/src/pyclass.rs:402]: Considering [the analogy with serde](https://serde.rs/container-attrs.html#rename_all), I would have expected to specify the convention as an argument to `rename_all`.

I understand that this mainly targets making enum variants matching the Python conventions, but I wonder whether we should give this a more specific name then, e.g. `rename_variants`?

File: tests/ui/invalid_pyclass_args.stderr
Comment (adamreichold) [inline tests/ui/invalid_pyclass_args.stderr:1]: I think we also need a Python test which applies this and checks the resulting variant names.

**Comment by DataTriny** [comment]
I'll try to update the `rename_all` attribute so that it accepts a case as argument and applies to struct fields as well, if there is interest in having this. Thanks for your feedback.

File: pyo3-macros-backend/src/pyclass.rs
Comment (DataTriny) [reply pyo3-macros-backend/src/pyclass.rs]: `rename_all` now behaves exactly like serde's one. Keeping the name as is.

**Comment by DataTriny** [comment]
@adamreichold Ready for another review.

File: pyo3-macros-backend/src/pyclass.rs
Comment (adamreichold) [inline pyo3-macros-backend/src/pyclass.rs:400]: Isn't `&format("{}", name)` equivalent to `&name.to_string()`?

File: pyo3-macros-backend/src/pymethod.rs
Comment (adamreichold) [inline pyo3-macros-backend/src/pymethod.rs:750]: I am not sure anything bad will ever happen here, but since we rely on zero-termination for memory safety, I would prefer if we did the renaming first and then added the zero termination. (Which could maybe use `.push('\0')` then?)

**Review by adamreichold — APPROVED** [review]
Thank you for going the extra mile and making this fully generic! I think this is now a powerful helper that can save quite a bit of boiler plate for many people.

File: pyo3-macros-backend/src/pymethod.rs
Comment (DataTriny) [reply pyo3-macros-backend/src/pymethod.rs]: I wasn't sure either so I tested and nothing bad happened. I even tried with reserved Rust keywords. Changing nonetheless so it's clearer.

**Comment by DataTriny** [comment]
codecov is failing again... I think I added enough tests and it can be ignored right?

**Comment by adamreichold** [comment]
> codecov is failing again... I think I added enough tests and it can be ignored right?

Yes, I think so too. The only reason I did not merge this yet is to give the other maintainers some time to chime in on this version of the code.

File: tests/test_class_attributes.rs
Comment (davidhewitt) [inline tests/test_class_attributes.rs:200]: It would be quite nice to have some kind of macro here which can test each type of rename_all in turn, so that we test they all work as expected.

```rust
test_case!(CamelCase, "camelCase", "ab_cd-ef", "abCdEf");
test_case!(SnakeCase, "snake_case", "ab_cd-ef", "ab_cd_ef");
```

etc

**Review by davidhewitt — APPROVED** [review]
Looks great to me, thanks for the solid implementation.

I just have one suggestion on an addition to the test suite (which will drive the coverage just a teeny bit higher too), otherwise I'm happy to see this merged whenever.

File: tests/test_class_attributes.rs
Comment (DataTriny) [reply tests/test_class_attributes.rs:200]: I added a macro which I think does the job. At least codecov likes it.

**Pull request merged**
